daemon/pipeline.c

```diff
@@ -0,0 +1,372 @@
+
+#include <glib.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <gst/gst.h>
+
+#include <common.h>
+#include <modules.h>
+#include <gdbus-util.h>
+#include <log.h>
+
+#include "dbus-interface.h"
+
+static MachinelearningServicePipeline *g_gdbus_instance = NULL;
+
+static int probe_pipeline (void *data);
+static void init_pipeline (void *data);
+static void exit_pipeline (void *data);
+
+static GHashTable *pipeline_table;
+
+typedef struct _pipeline {
+  GstElement *element;
+  gint64 id;
+  GMutex lock;
+  gchar *description;
+  // Is this enough?
+  // Is register any gst related callback valid?
+  // When there occur error or state change -> should broadcast it via d-bus to clients
+} pipeline_s;
+
+// See ml_pipeline_destroy
+static void _pipeline_free (gpointer data)
+{
+  pipeline_s *p;
+
+  if (!data) {
+    g_critical ("internal error, the data should not be NULL");
```

Comment:
This stops the daemon. Is it ok? (is it supposed to be an internal error that cannot be invoked by user s' or applications' behaviors/inputs?)